### PR TITLE
[Dashboard] Remove opencensus from agent proxy export

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR removes the Opencensus usage on proxy export. Previously, OpenCensus APIs we are using for proxy export deepcopies the whole data {labels -> data} whenever there's a new export which causes O(N^2) write on metrics record. See the below section for more details on removing Opencensus. 

Instead of using their APIs, we will store the aggregation data in memory and export them using a custom Prometheus exporter (0 deepcopies, purely done by lock). Below is the flamegraph for the same workload (100 actors + submitting 1000 tasks per second + 1 second metrics export). Before this fix, the CPU usage was > 100% all the time. With this fix, the CPU usage is only about 10~15% with the same workload. 

Before (CPU > 100% all the time)
Screen Shot 2022-11-18 at 4 16 13 AM

After (CPU 10~15%)
Screen Shot 2022-11-18 at 4 16 06 AM

## Removing OpenCensus

Opencensus has caused lots of issues from metrics agent due to 
1. We are using their API wrong. But API is not well documented, and their official API cannot be used for proxy export.
2. inefficient implementation that uses deep copy all the time which caused the agent memory leak. 
3. Their official API is not well documented and the project is not maintained anymore (it is migrated to OpenTelemetry)

This PR is the first step to remove OpenCensus from the dashboard agent. Note that one possible direction is to migrate to OpenTelemetry, but the project still looks like it needs more time (e.g., their doc is still pretty bad). So it is worth waiting longer.

This means we will only support Prometheus format. But it is not a regression + it seems the same for other systems like K8s.

There will be 3 PRs. 
- (current) Remove Opencesnsus usage from proxy export
- Remove opencensus usage for reporter_agent (and instead use promtheus client directly)
- Clean up & aggregating worker ids (not sure if we should do it, but it doesn't look too hard)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/metrics_agent.py [hunk 1]
```diff
@@ -66,28 +72,303 @@ def name(self):
...
+            label_keys: The fixed label keys of the metric.
+            metric_units: Units of the metric.
+            label_values: The values of `label_keys`.
+            agg_data: `opencensus.stats.aggregation_data.AggregationData` object.
+                Aggregated data that needs to be converted as Prometheus samples
+
+        Returns:
+            A Prometheus metric object
+        """
+        assert self._components_lock.locked()
```
Comment (rkooo567) [inline python/ray/_private/metrics_agent.py:280]: this method is same as `to_metric` inside Prometheus_exporter.py

**Comment by rkooo567** [comment]
Running the agent stress test with export frequency 1000ms

File: python/ray/_private/metrics_agent.py [hunk 2]
```diff
@@ -66,28 +72,303 @@ def name(self):
...
+            if name not in self._metrics:
+                self._metrics[name] = OpencensusProxyMetric(
+                    name, descriptor.description, descriptor.unit, label_keys
+                )
+            self._metrics[name].record(metric)
+
+
+class OpenCensusProxyCollector:
+    def __init__(self, namespace: str, component_timeout_s: int = 60):
+        """Promehtue collector implementation for opencensus proxy export.
```
Comment (ericl) [inline python/ray/_private/metrics_agent.py:190]:
```suggestion
        """Prometheus collector implementation for opencensus proxy export.
```

File: python/ray/_private/metrics_agent.py [hunk 3]
```diff
@@ -66,28 +72,303 @@ def name(self):
...
+            self._components[key].record(metrics)
 
-    def put_tag_vals(self, view_name: str, tag_vals: Tuple[str]):
-        self._view_to_owned_tag_vals[view_name].add(tag_vals)
+    def clean_stale_components(self):
+        """Clean up stale components.
+
+        Stale means the component is dead or unresponsive.
+
+        Stale components won't be reported to Promemetheus anymore.
```
Comment (ericl) [inline python/ray/_private/metrics_agent.py:235]:
```suggestion
        Stale components won't be reported to Prometheus anymore.
```

**Review by ericl — APPROVED** [review]
Nice, happy to see this strategy worked out.

**Comment by rkooo567** [comment]
Memory usage before: (reached 2GB in 15m)

Screen Shot 2022-11-18 at 6 06 43 PM

Memory usage after (112MB after 20m):
Screen Shot 2022-11-18 at 6 31 31 PM

**Pull request merged**